presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -276,59 +282,74 @@ public IPrestoSparkTaskExecutor doCreate(
                 false);
 
-        OutputBufferMemoryManager memoryManager = new OutputBufferMemoryManager(
-                sinkMaxBufferSize.toBytes(),
-                () -> queryContext.getTaskContextByTaskId(taskId).localSystemMemoryContext(),
-                notificationExecutor);
-        PrestoSparkRowBuffer rowBuffer = new PrestoSparkRowBuffer(memoryManager);
-
-        ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkMutableRow>> shuffleInputs = ImmutableMap.builder();
-        ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkSerializedPage>> broadcastInputs = ImmutableMap.builder();
+        ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkMutableRow>> rowInputs = ImmutableMap.builder();
+        ImmutableMap.Builder<PlanNodeId, Iterator<PrestoSparkSerializedPage>> pageInputs = ImmutableMap.builder();
         for (RemoteSourceNode remoteSource : fragment.getRemoteSourceNodes()) {
-            List<Iterator<PrestoSparkMutableRow>> shuffleRemoteSourceInputs = new ArrayList<>();
-            List<Iterator<PrestoSparkSerializedPage>> broadcastRemoteSourceInputs = new ArrayList<>();
+            List<Iterator<PrestoSparkMutableRow>> remoteSourceRowInputs = new ArrayList<>();
+            List<Iterator<PrestoSparkSerializedPage>> remoteSourcePageInputs = new ArrayList<>();
             for (PlanFragmentId sourceFragmentId : remoteSource.getSourceFragmentIds()) {
                 Iterator<Tuple2<MutablePartitionId, PrestoSparkMutableRow>> shuffleInput = inputs.getShuffleInputs().get(sourceFragmentId.toString());
                 Broadcast<List<PrestoSparkSerializedPage>> broadcastInput = inputs.getBroadcastInputs().get(sourceFragmentId.toString());
-                checkArgument(shuffleInput != null || broadcastInput != null, "Input not found for sourceFragmentId: %s", sourceFragmentId);
-                checkArgument(shuffleInput == null || broadcastInput == null, "Single remote source cannot accept both, broadcast and shuffle inputs");
+                List<PrestoSparkSerializedPage> inMemoryInput = inputs.getInMemoryInputs().get(sourceFragmentId.toString());
+
                 if (shuffleInput != null) {
-                    shuffleRemoteSourceInputs.add(Iterators.transform(shuffleInput, tuple -> tuple._2));
+                    checkArgument(broadcastInput == null, "single remote source is not expected to accept different kind of inputs");
+                    checkArgument(inMemoryInput == null, "single remote source is not expected to accept different kind of inputs");
+                    remoteSourceRowInputs.add(Iterators.transform(shuffleInput, tuple -> tuple._2));
+                    continue;
                 }
+
                 if (broadcastInput != null) {
+                    checkArgument(inMemoryInput == null, "single remote source is not expected to accept different kind of inputs");
                     // TODO: Enable NullifyingIterator once migrated to one task per JVM model
```

Comment:
Is this comment still relevant?

Reply:
Yes, it is still relevant. Transitioning from rows to pages saves memory on the driver. `NullifyingIterator` is needed to save memory on executors.